File: languages/r/web3_client.R
#!/usr/bin/env Rscript

library(httr)
library(jsonlite)

RPC_URL <- "https://mainnet.infura.io/v3/YOUR_KEY"

rpc_call <- function(method, params = list()) {
  body <- list(
    jsonrpc = "2.0",
    id = 1,
    method = method,
    params = params
  )

  response <- POST(
    RPC_URL,
    body = toJSON(body, auto_unbox = TRUE),
    content_type_json()
  )

  content <- content(response, as = "parsed", type = "application/json")
  return(content$result)
}

hex_to_number <- function(hex) {
  hex_clean <- sub("^0x", "", hex)
  strtoi(hex_clean, base = 16L)
}

get_block_number <- function() {
  result <- rpc_call("eth_blockNumber")
  return(hex_to_number(result))
}

get_balance <- function(address) {
  result <- rpc_call("eth_getBalance", list(address, "latest"))
  wei <- hex_to_number(result)
  return(as.numeric(wei) / 1e18)
}

get_gas_price <- function() {
  result <- rpc_call("eth_gasPrice")
  wei <- hex_to_number(result)
  return(as.numeric(wei) / 1e9)
}

check_multiple_balances <- function(addresses) {
  balances <- sapply(addresses, get_balance)
  data.frame(
    address = addresses,
    balance = balances,
    stringsAsFactors = FALSE
  )
}

# Main
cat("Current block:", get_block_number(), "\n")

vitalik <- "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"
balance <- get_balance(vitalik)
cat(sprintf("Vitalik balance: %.4f ETH\n", balance))

gas_price <- get_gas_price()
cat(sprintf("Gas price: %.2f Gwei\n", gas_price))

# Check multiple balances
addresses <- c(
  "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
  "0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb",
  "0xBE0eB53F46cd790Cd13851d5EFf43D12404d33E8"
)

balances_df <- check_multiple_balances(addresses)
print(balances_df)

# Statistical analysis
cat(sprintf("\nAverage balance: %.4f ETH\n", mean(balances_df$balance)))
cat(sprintf("Total balance: %.4f ETH\n", sum(balances_df$balance)))

File: languages/julia/web3_client.jl
using HTTP
using JSON3

const RPC_URL = "https://mainnet.infura.io/v3/YOUR_KEY"

function rpc_call(method::String, params::Vector=[])
    body = Dict(
        "jsonrpc" => "2.0",
        "id" => 1,
        "method" => method,
        "params" => params
    )

    headers = ["Content-Type" => "application/json"]
    response = HTTP.post(RPC_URL, headers, JSON3.write(body))

    result = JSON3.read(String(response.body))
    return result.result
end

function hex_to_number(hex::String)
    hex_clean = replace(hex, "0x" => "")
    return parse(Int64, hex_clean, base=16)
end

function get_block_number()
    result = rpc_call("eth_blockNumber")
    return hex_to_number(result)
end

function get_balance(address::String)
    result = rpc_call("eth_getBalance", [address, "latest"])
    wei = hex_to_number(result)
    return wei / 1e18
end

function get_gas_price()
    result = rpc_call("eth_gasPrice")
    wei = hex_to_number(result)
    return wei / 1e9
end

function check_multiple_balances(addresses::Vector{String})
    balances = Dict{String, Float64}()

    @sync for address in addresses
        @async begin
            balance = get_balance(address)
            balances[address] = balance
        end
    end

    return balances
end

# Main
println("Current block: ", get_block_number())

vitalik = "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"
balance = get_balance(vitalik)
println("Vitalik balance: ", round(balance, digits=4), " ETH")

gas_price = get_gas_price()
println("Gas price: ", round(gas_price, digits=2), " Gwei")

# Check multiple balances
addresses = [
    "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
    "0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb",
    "0xBE0eB53F46cd790Cd13851d5EFf43D12404d33E8"
]

balances = check_multiple_balances(addresses)
println("\nBalances:")
for (addr, bal) in balances
    println("$addr: $(round(bal, digits=4)) ETH")
end

# Statistical analysis
all_balances = collect(values(balances))
println("\nStatistics:")
println("Average: ", round(sum(all_balances) / length(all_balances), digits=4), " ETH")
println("Total: ", round(sum(all_balances), digits=4), " ETH")
println("Max: ", round(maximum(all_balances), digits=4), " ETH")
